### ISSUE TITLE
Daemon not responding yet net.OpError - Connection refused

### ISSUE THREAD

**Issue by jeevanullas** [docker/machine#1111]
Hey guys,

Great work here. Today I looked at the amazonec2 driver and used it against Eucalyptus. It works without any problem, I had to modify the driver to not have VPC mandatory because Eucalyptus currently does not have VPC support. 

The thing that seems weird is when I create a machine it seems to always hung up on 

Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)

I have to Ctrl+C this and if I then check it seems to be reportedly running just fine (not sure if it's ACTIVE though)

$ docker-machine_linux-amd64 ls
NAME                      ACTIVE   DRIVER      STATE     URL                         SWARM
docker-host-euca-jeevan            amazonec2   Running   tcp://192.168.1.2:2376 

The complete debug output along with the parameters is shown below:

$ docker-machine_linux-amd64 -D create --driver amazonec2 --amazonec2-ami emi-544780e1 --amazonec2-access-key  --amazonec2-secret-key  --amazonec2-zone emea-az-01 --amazonec2-region eucalyptus --amazonec2-instance-type m1.large docker-host-euca-jeevan

Launching instance...
creating key pair: docker-host-euca-jeevan
configuring security group
found existing security group (docker-machine)
configuring security group authorization for 0.0.0.0/0
launching instance
waiting for ip address to become available
Got the IP Address, it's "10.30.4.131"
created instance ID i-60771744, IP address 10.30.4.131, Private IP address 
Settings tags for instance
Getting to WaitForSSH function...

generating server cert: /home/jeevanullas/.docker/machine/machines/docker-host-euca-jeevan/server.pem ca-key=/home/jeevanullas/.docker/machine/certs/ca.pem private-key=/home/jeevanullas/.docker/machine/certs/ca-key.pem org=docker-host-euca-jeevan

Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)
Daemon not responding yet: %!(EXTRA *net.OpError=dial tcp 192.168.1.2:2376: connection refused)

I could use the SSH private key and login to the instance. I found docker to be running and listening correctly on the right port. I tried doing a telnet from the same host where I created the machine from and it worked for port 2376.

I checked the security group and all looks good there as well.

GROUP	sg-402dda41	849254641345	docker-machine	Docker+Machine	
PERMISSION	849254641345	docker-machine	ALLOWS	tcp	22	22	FROM	CIDR	0.0.0.0/0	ingress
PERMISSION	849254641345	docker-machine	ALLOWS	tcp	2376	2376	FROM	CIDR	0.0.0.0/0	ingress

$ ssh -i id_rsa ubuntu@192.168.1.2
The authenticity of host '192.168.1.2 (192.168.1.2)' can't be established.
ECDSA key fingerprint is 9a:85:76:69:c3:82:c8:d8:f3:71:24:24:fa:ae:55:ad.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '192.168.1.2' (ECDSA) to the list of known hosts.
Welcome to Ubuntu 14.04.2 LTS (GNU/Linux 3.13.0-49-generic x86_64)

 * Documentation:  https://help.ubuntu.com/

  System information as of Tue May  5 10:48:05 UTC 2015

  System load:  0.9               Processes:              116
  Usage of /:   65.7% of 2.11GB   Users logged in:        1
  Memory usage: 22%               IP address for eth0:    10.30.4.131
  Swap usage:   0%                IP address for docker0: 172.17.42.1

  Graph this data and manage this system at:
    https://landscape.canonical.com/

  Get cloud support with Ubuntu Advantage Cloud Guest:
    http://www.ubuntu.com/business/services/cloud

*** System restart required ***
Last login: Tue May  5 10:45:17 2015 from 

ubuntu@docker-host-euca-jeevan:~$ sudo su -

root@docker-host-euca-jeevan:~# netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1051/sshd       
tcp6       0      0 :::22                   :::*                    LISTEN      1051/sshd       
tcp6       0      0 :::2376                 :::*                    LISTEN      29947/docker    
udp        0      0 0.0.0.0:36472           0.0.0.0:*                           557/dhclient    
udp        0      0 0.0.0.0:68              0.0.0.0:*                           557/dhclient    
udp6       0      0 :::34284                :::*                                557/dhclient    

root@docker-host-euca-jeevan:~# ps aux | grep docker
root     29947  0.1  2.0 257732 10248 ?        Ssl  10:48   0:00 /usr/bin/docker -d -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock --storage-driver aufs --tlsverify --tlscacert /etc/docker/ca.pem --tlscert /etc/docker/server.pem --tlskey /etc/docker/server-key.pem --label provider=amazonec2
root     30095  0.0  0.1  10460   920 pts/2    S+   10:51   0:00 grep --color=auto docker

root@docker-host-euca-jeevan:~# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

root@docker-host-euca-jeevan:~# docker version
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64

$ telnet 192.168.1.2 2376
Trying 192.168.1.2...
Connected to 192.168.1.2.
Escape character is '^]'.
quit
Connection closed by foreign host.

Is it because the instance needs a reboot or the service is listening on an IPv6 address? Any thoughts on how to proceed further are really appreciated. 

Version of docker machine:
$ docker-machine_linux-amd64 -v
docker-machine_linux-amd64 version 0.2.0 (e4b6ff1)

Cheers,
Deependra

**Comment by jeevanullas** [comment]
Docker daemon log file on the instance looks good as well

cat /var/log/upstart/docker.log 
Waiting for /var/run/docker.sock
INFO[0000] +job init_networkdriver()                    
INFO[0000] +job serveapi(unix:///var/run/docker.sock)   
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
/var/run/docker.sock is up
INFO[0000] -job init_networkdriver() = OK (0)           
WARN[0000] Your kernel does not support cgroup swap limit. 
INFO[0000] Loading containers: start.                   

INFO[0000] Loading containers: done.                    
INFO[0000] docker daemon: 1.6.0 4749651; execdriver: native-0.2; graphdriver: aufs 
INFO[0000] +job acceptconnections()                     
INFO[0000] -job acceptconnections() = OK (0)            
INFO[0000] Daemon has completed initialization          
INFO[0004] GET /v1.18/version                           
INFO[0004] +job version()                               
INFO[0004] -job version() = OK (0)                      
INFO[0008] GET /v1.18/version                           
INFO[0008] +job version()                               
INFO[0008] -job version() = OK (0)                      
INFO[0015] Received signal 'terminated', starting shutdown of docker... 
INFO[0015] -job serveapi(unix:///var/run/docker.sock) = OK (0) 
INFO[0000] +job init_networkdriver()                    
INFO[0000] +job serveapi(tcp://0.0.0.0:2376, unix:///var/run/docker.sock) 
INFO[0000] Listening for HTTP on tcp (0.0.0.0:2376)     
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
INFO[0000] -job init_networkdriver() = OK (0)           
WARN[0000] Your kernel does not support cgroup swap limit. 
INFO[0000] Loading containers: start.                   

INFO[0000] Loading containers: done.                    
INFO[0000] docker daemon: 1.6.0 4749651; execdriver: native-0.2; graphdriver: aufs 
INFO[0000] +job acceptconnections()                     
INFO[0000] -job acceptconnections() = OK (0)            
INFO[0000] Daemon has completed initialization          
INFO[0112] Received signal 'terminated', starting shutdown of docker... 
INFO[0112] -job serveapi(tcp://0.0.0.0:2376, unix:///var/run/docker.sock) = OK (0) 
INFO[0000] +job serveapi(tcp://0.0.0.0:2376, unix:///var/run/docker.sock) 
INFO[0000] Listening for HTTP on tcp (0.0.0.0:2376)     
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
INFO[0000] +job init_networkdriver()                    
INFO[0000] -job init_networkdriver() = OK (0)           
WARN[0001] Your kernel does not support cgroup swap limit. 
INFO[0001] Loading containers: start.                   

INFO[0001] Loading containers: done.                    
INFO[0001] docker daemon: 1.6.0 4749651; execdriver: native-0.2; graphdriver: aufs 
INFO[0001] +job acceptconnections()                     
INFO[0001] -job acceptconnections() = OK (0)            
INFO[0001] Daemon has completed initialization  

I tried by rebooting the instance but no luck.

**Comment by nathanleclaire** [comment]
The issue seems to be that Machine is attempting to dial `192.168.1.2:2376`, instead of `10.30.4.131` (the public IP address).  I wonder why.

**Comment by jeevanullas** [comment]
Actually, 192.168.1.2 is the public IP of the instance in this case.

**Comment by jeevanullas** [comment]
Coming back at it. I modified the placement of this line here:

https://github.com/docker/machine/blob/master/drivers/amazonec2/amazonec2.go#L399

And moved it up a bit before the IP address check kicks in. This way we always get the public IP address , note that in Eucalyptus when the instance is kicked off it takes at-least 5 seconds before a public IP gets on to it.

But I still continue to face issue with the connection. I found out that telnet does not work. My bad. I did telnet earlier from the same host as the instance and did work but if I try it out from outside the instance it does not work.

This brought me back to ipv6 issue. It appears that the docker daemon listen on tcp6 but can accept connection on ipv4 and ipv6 both (netstat output can be misleading?). Some useful links I found but got no further help on this and I am still stuck:

http://gesellix.net/docker-and-ipv6-on-ubuntu-13-10-saucy-salamander/
https://github.com/docker/docker/issues/2174

Hope this is useful information. Let me know if you want me to contribute the pull request as mentioned above for the IP address thing. Cheers,
